### PR TITLE
WIP: Gitlab support update

### PR DESCRIPTION
**Pull request by ge-o-rg**
#### Submitter checklist

- [ ] Change is code complete and matches issue description.
- [ ] Change is covered by existing or new tests.

#### Description
Updated GitLab support and fixing issues 1932 and 3339
This is work in progress but I need to build an image file for jx to pull and test.

#### Which issue this PR fixes

fixes #

### CODE REVIEW

**Comment by dodwyer** [comment]
Looking forward to seeing these changes move forward. Happy to test across a number of configurations we have

**Comment by ge-o-rg** [comment]
> Looking forward to seeing these changes move forward. Happy to test across a number of configurations we have

Should hopefully be done in next few days (famous last words :)). There's a couple more things I need to check / make sure they work as intended. The PR allows me to get image done so I can cross check the changes in the jx build pipeline.

File: pkg/gits/gitlab.go [hunk 1]
```diff
@@ -691,3 +703,39 @@ func (g *GitlabProvider) ListCommits(owner, repo string, opt *ListCommitsArgumen
...
+		ListOptions: gitlab.ListOptions{
+			Page:    0,
+			PerPage: pageSize,
+		},
+		SourceBranch: &branchName,
+	}
+	mergeRequests, _, err := g.Client.MergeRequests.ListProjectMergeRequests(pid, opt)
+
+	if err != nil {
+		return 0, "", err
```
Comment (ccojocar) [inline pkg/gits/gitlab.go]: ditto

**Comment by ccojocar** [comment]
@ge-o-rg What's left to be done? Thanks

**Comment by ge-o-rg** [comment]
@ccojocar Things that need to be done:

* need to finish tests
* test it on github, bitbucket etc 
* review your proposed changes 
*  for gitlab we  currently only install a  push event web hook but not for MR opened. There’s a MR web hook we can trigger but this also triggers when MR gets updated (same as push commit) or when MR gets merged so trying to figure out best way to control which requests to listen to server side.  I found this plugin for Jenkins that might help:https://github.com/jenkinsci/generic-webhook-trigger-plugin  and Will pointed out this one: https://github.com/Argelbargel/gitlab-branch-source-plugin
So just need to figure out best way forward on this for Jenkins static, eg future proof it. If you have any suggests / comments then super appreciated.

The actual functionality does work on my gitlab test setup, as in: I’m able to created MRs and push commits, this triggers a Build and generates a Preview if enabled and posts the link as comment and on merge it gets promoted to staging without any errors.

was keen on getting this done by last Sunday but urged and sudden work commitments meant I wasn’t able to.

**Comment by ccojocar** [comment]
>for gitlab we currently only install a push event web hook but not for MR opened. There’s a MR web hook we can trigger but this also triggers when MR gets updated (same as push commit) or when MR gets merged so trying to figure out best way to control which requests to listen to server side. I found this plugin for Jenkins that might help:https://github.com/jenkinsci/generic-webhook-trigger-plugin and Will pointed out this one: https://github.com/Argelbargel/gitlab-branch-source-plugin
So just need to figure out best way forward on this for Jenkins static, eg future proof it. If you have any suggests / comments then super appreciated.

We should probably adjust the `CreateWebhook` options in Gitlab provider:

https://github.com/jenkins-x/jx/blob/ef1006c4d4f9a21fa0de2a55ae4cbb4083374c9e/pkg/gits/gitlab.go#L454)

I am not sure if some additional configuration is required for classic Jenkins.

File: pkg/gits/gitlab.go [hunk 2]
```diff
@@ -691,3 +703,39 @@ func (g *GitlabProvider) ListCommits(owner, repo string, opt *ListCommitsArgumen
 func (g *GitlabProvider) AddLabelsToIssue(owner, repo string, number int, labels []string) error {
 	return fmt.Errorf("Getting content not supported on gitlab yet")
 }
+
+// GetPRNumFromBranchName returns the PR Number from a branch name.
+func (g *GitlabProvider) GetPRNumFromBranchName(owner string, repo *GitRepository, branchName string) (int, string,  error) {
+
+	pid, err := g.projectId(owner, g.Username, repo.Name)
+	if err != nil {
+		return 0, "", err
```
[the remark this answers is not in the document]
Comment (ge-o-rg) [reply pkg/gits/gitlab.go]: Has been sorted.

File: pkg/gits/gitlab.go (hunk 1 above)
Comment (ge-o-rg) [reply pkg/gits/gitlab.go]: Has been sorted.

**Comment by hferentschik** [comment]
@ge-o-rg, hi, how it is going? Just wanted to touch base to hear what the state of this pull request is? Is this still something you are working on?

**Comment by hferentschik** [comment]
I am going to close this pull request for now. If you are going to work on it and get it rebased feel free to open a new pull request.

**Comment by ge-o-rg** [comment]
Hello!
Apologies for the late reply, i'm in the final staging of doing testing on
my end should hopefully be able to update it by end of the week.
If you could reopen that would be great cause i need push to PR to be able
to create images and test my changes.
Thanks!

**Pull request closed without merging**